lib/vanagon/project.rb

```diff
@@ -82,6 +82,32 @@ def get_files
       files.flatten.uniq
     end
 
+    # Returns a filtered out set of components only including those
+    # components necessary to build a specific component. This is a
+    # recursive function that will call itself until it gets to a
+    # component with no build requirements
+    #
+    # @param name [String] name of component to add. must be present in configdir/components and named $name.rb currently
+    def filter_component(name)
+      components_to_include = []
+      filter_component = get_component(name)
+      return nil if filter_component.nil?
+      components_to_include << filter_component
+      filter_component.build_requires.each do |build_requirement|
+        unless get_component(build_requirement).nil?
+          components_to_include += filter_component(build_requirement)
+        end
+      end unless filter_component.build_requires.nil?
+      components_to_include
+    end
```

Comment:
OK, I dug into `filter_component` a little bit and rewrote it to understand it. The important takeaways are: don't shadow variable names by reusing method names, and consider streamlining some of your operations a little bit by standardizing on the expected values are expected from outside your new code. Comments inline:


```ruby
def filter_component(name)
  filtered_component = get_component(name)
  return nil unless filtered_component

  included_components = [filtered_component]

  # is the default value of #build_requires `nil` or `false`?
  # if it's iteratable, shouldn't it just be []? So we'd test
  # for whether or not it's empty, not if it's nil. If it's
  # not [], then we should probably fix and formalize that.
  # here, I'm testing to see if it's #empty? but that's kind of
  # a sign that it's a wonky interface more than anything else.
  unless filtered_component.build_requires.empty?
    filtered_component.build_requires.each do |build_requirement|
      unless get_component(build_requirement).empty?
        included_components << filtered_component(build_requirement)
      end
    end
  end
  included_components
end

# I would allow this method to return an empty array rather than coercing it
# to nil and then I would standardize on that as an expected interface since
# iterating over an empty array is functionally a NOOP.
# 
# >> [].each { |v| puts v }
# => []
# >>
# 
def get_component(name)
  @components.select { |comp| comp.name.to_s == name.to_s }
end
```

This is probably broken or missing pieces since I don't have good data to feed to it here to test this out. But the idea is that if something is expected to be iterable, let it be returned as an empty iterable (basically anything inheriting from `Enumerable`). If it's imperative that you not attempt to operate on an empty `Enumerable`, then use `empty?` instead of `nil?`.